### PR TITLE
[BE][DTensor] validate the mesh argument in DeviceMesh construction

### PR DESCRIPTION
**Pull request by XilunWu**
## What's in this PR
DeviceMesh's __init__ function now requires all calling ranks to pass the same `mesh` argument.

## Why
We want to enforce SPMD style of programs using DTensor. Before this PR, 2-D Parallel API (e.g. _create_1d_device_mesh) defines different DeviceMesh on different ranks. After this PR, it defines each sub-meshes and simply perform communications on the one that it is associated with.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99094

Differential Revision: [D45165511](https://our.internmc.facebook.com/intern/diff/D45165511)

### CODE REVIEW

**Comment by XilunWu** [comment]
TODO: need to change existing TP tests

File: test/distributed/_tensor/test_device_mesh.py [hunk 1]
```diff
@@ -94,6 +94,17 @@ def test_lazy_init_device_mesh(self):
         else:
             assert mesh.get_coordinate() is None
 
+    @with_comms
+    def test_validate_device_mesh(self):
+        mesh = torch.arange(self.world_size).reshape(2, -1)
+        mesh_subgp_1 = mesh[0]
```
Comment (wanchaol) [inline test/distributed/_tensor/test_device_mesh.py:100]:
```suggestion
        mesh_subpg_1 = mesh[0]
```

File: test/distributed/_tensor/test_device_mesh.py [hunk 2]
```diff
@@ -94,6 +94,17 @@ def test_lazy_init_device_mesh(self):
...
+        mesh_subgp_2 = mesh[1]
```
Comment (wanchaol) [inline test/distributed/_tensor/test_device_mesh.py:101]:
```suggestion
        mesh_subpg_2 = mesh[1]
```

**Review by wanchaol — APPROVED** [review]
Looks good! Thanks for working on it!

Left a small nit. Could you import this to internal too to make sure nothing breaks

**Comment by XilunWu** [comment]
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45165511).

**Pull request closed without merging**